From 90e09aa62c98d1f782343b1cf5cecb890ebe2e0c Mon Sep 17 00:00:00 2001
From: Panu Matilainen <pmatilai@redhat.com>
Date: Fri, 10 May 2024 14:52:37 +0300
Subject: [PATCH] Always create %specpartsdir on build

There's no reason %specpartsdir should be dependent on %setup use,
just create it when we create the build directory too. Update tests
accordingly.

The spec parse test is noteworthy, the specparts dir creation no longer
shows up in spec parse output, which certainly is the way it should be:
this is rpm internal infrastructure stuff and nothing to do with spec
*parse*, this all only takes place during builds.

Fixes: #3063
---
 build/build.c     |  1 +
 build/parsePrep.c | 10 ----------
 tests/rpmbuild.at |  1 +
 tests/rpmspec.at  |  2 --
 4 files changed, 2 insertions(+), 12 deletions(-)

diff --git a/build/build.c b/build/build.c
index fd0e4be3f0..0c2d738621 100644
--- a/build/build.c
+++ b/build/build.c
@@ -325,6 +325,7 @@ static rpmRC doBuildDir(rpmSpec spec, int test, StringBuf *sbp)
 			   "%{_fixperms} '", spec->buildDir, "'\n",
 			   "%{__rm} -rf '", spec->buildDir, "'\n",
 			   "%{__mkdir_p} '", spec->buildDir, "'\n",
+			   "%{__mkdir_p} '%{specpartsdir}'\n",
 			   NULL);
 
     rpmRC rc = doScript(spec, RPMBUILD_MKBUILDDIR, "%mkbuilddir",
diff --git a/build/parsePrep.c b/build/parsePrep.c
index 987a6d990c..4754827f56 100644
--- a/build/parsePrep.c
+++ b/build/parsePrep.c
@@ -276,16 +276,6 @@ void doSetupMacro(rpmMacroBuf mb, rpmMacroEntry me, ARGV_t margs, size_t *parsed
 	free(buf);
     }
 
-    /* mkdir for dynamic specparts */
-    if (rpmMacroIsDefined(spec->macros, "specpartsdir")) {
-	buf = rpmExpand("rm -rf '%{specpartsdir}'", NULL);
-	appendMb(mb, buf, 1);
-	free(buf);
-	buf = rpmExpand("%{__mkdir_p} '%{specpartsdir}'", NULL);
-	appendMb(mb, buf, 1);
-	free(buf);
-    }
-
     appendMb(mb, getStringBuf(after), 0);
 
     /* Fix the permissions of the setup build tree */
diff --git a/tests/rpmbuild.at b/tests/rpmbuild.at
index 61264f5ac4..dd2d017bdc 100644
--- a/tests/rpmbuild.at
+++ b/tests/rpmbuild.at
@@ -278,6 +278,7 @@ runroot_other find /build/BUILD|sort
 /build/BUILD/simple-1.0-build/BUILDROOT/opt
 /build/BUILD/simple-1.0-build/BUILDROOT/opt/bin
 /build/BUILD/simple-1.0-build/BUILDROOT/opt/bin/simple
+/build/BUILD/simple-1.0-build/SPECPARTS
 /build/BUILD/simple-1.0-build/simple
 ],
 [ignore])
diff --git a/tests/rpmspec.at b/tests/rpmspec.at
index ed5872aaf6..0de41f3ea2 100644
--- a/tests/rpmspec.at
+++ b/tests/rpmspec.at
@@ -334,8 +334,6 @@ if [ $STATUS -ne 0 ]; then
   exit $STATUS
 fi
 cd 'hello-1.0'
-rm -rf '/build/BUILD/hello-1.0-build/SPECPARTS'
-/usr/bin/mkdir -p '/build/BUILD/hello-1.0-build/SPECPARTS'
 /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
 
 echo "Patch #0 (hello-1.0-modernize.patch):"
